File: Gillian-JS/Examples/JaVerT/BST.js
"use strict";

/**
	@pred nounfold NullableObject(o) :
		types (o : Obj),
		types (o : Null);

	@pred Node(+n : Obj, val : Num, left, right) :
		JSObject(n) *
		DataProp(n, "value", val) * DataProp(n, "left",  left) * DataProp(n, "right", right);

	@pred BST(+n, K : Set) :
		(n == null) * (K == -{ }-),

		Node(n, #val, #left, #right) * BST(#left, #KL) * BST(#right, #KR) *
		(K == -u- (#KL, -{ #val }-, #KR)) *
		(forall #x : Num. ((! (#x --e-- #KL)) \/ (#x <# #val))) *
		(forall #x : Num. ((! (#x --e-- #KR)) \/ (#val <# #x))) *
        types(#val : Num, #KL : Set, #KR : Set);
*/

/**
	@id makeNode

	@pre
		(v == #v) * types (#v : Num)

	@post
		Node(ret, #v, null, null)
*/
function make_node(v)
{
  var node = {
    value : v,
    left  : null,
    right : null
  };
  return node;
}

/**
	@id insert

    @pre
        (v == #v) * (t == #t) *
		GlobalObject() * ObjectPrototype($lobj_proto) *
		BST(#t, #K) * types (#v : Num) *
		scope(make_node : #makeNode) * JSFunctionObject(#makeNode, "makeNode", _, _, _) *
		scope(insert : #insert) * JSFunctionObject(#insert, "insert", _, _, _)

	@post
		GlobalObject() * ObjectPrototype($lobj_proto) *
		BST(ret, -u- (#K, -{ #v }-)) * types (ret : Obj) *
		scope(make_node : #makeNode) * JSFunctionObject(#makeNode, "makeNode", _, _, _) *
		scope(insert : #insert) * JSFunctionObject(#insert, "insert", _, _, _)
*/
function insert(v, t)
{
  var result;

  if (t === null) {
  	return make_node(v);
  }

  if (v < t.value)
    t.left = insert(v, t.left);
  else if (v > t.value)
    t.right = insert(v, t.right);

  return t;
}

/**
	@id find

	@pre
		(t == #t) * BST(#t, #K) * (v == #v) * types (#v : Num) *
		scope(find : #find) * JSFunctionObject(#find, "find", _, _, _) *
		GlobalObject() * ObjectPrototype($lobj_proto)

	@post
		BST(#t, #K) * (ret == (#v -e- #K)) *
		scope(find : #find) * JSFunctionObject(#find, "find", _, _, _) *
		GlobalObject() * ObjectPrototype($lobj_proto)
*/
function find (v, t)
{
	var result;

	if (t === null)
		result = false;
	else if (v === t.value)
		result = true;
	else {
		if (v < t.value)
		  result = find(v, t.left)
		else
		  result = find(v, t.right);
	}

	return result;
}

/**
	@id findMin

	@pre
		GlobalObject() * ObjectPrototype($lobj_proto) *
		(t == #t) * BST(#t, #K) * types(#t : Obj) *
		scope(find_min : #findMin) * JSFunctionObject(#findMin, "findMin", _, _, _)

	@post
		GlobalObject() * ObjectPrototype($lobj_proto) *
		BST(#t, #K) * (ret --e-- #K) * types(ret : Num) *
		(forall #x : Num. ((! (#x --e-- #K)) \/ (ret <=# #x))) *
		scope(find_min : #findMin) * JSFunctionObject(#findMin, "findMin", _, _, _)
*/
function find_min(t)
{
	var result;

	if (t.left === null)
		result = t.value;
	else
		result = find_min(t.left);

	return result;
}

/**
	@id remove

	@pre
		GlobalObject() * ObjectPrototype($lobj_proto) *
		(t == #t) * BST(#t, #K) *
		(v == #v) * types (#v : Num) *
		scope(remove : #remove) * JSFunctionObject(#remove, "remove", _, _, _) *
		scope(find_min : #findMin) * JSFunctionObject(#findMin, "findMin", _, _, _)

	@post
		GlobalObject() * ObjectPrototype($lobj_proto) *
		BST(ret, #K_new) * (#K_new == #K -d- -{ #v }-) * NullableObject(ret) *
		scope(remove : #remove) * JSFunctionObject(#remove, "remove", _, _, _) *
		scope(find_min : #findMin) * JSFunctionObject(#findMin, "findMin", _, _, _)
*/
function remove(v, t)
{
	if (t === null)
		return null;

	if (v === t.value) {
		if (t.left === null) {
				return t.right;
			}
        else
		if (t.right === null) {
	  			return t.left;
			}
		else {
			var min = find_min(t.right);
			t.right = remove(min, t.right);
			t.value = min;
		}
	}
    else if (v < t.value)
		t.left = remove(v, t.left);
    else
		t.right = remove(v, t.right);

  return t;
}

File: Gillian-C/examples/verification/bst.c
#include <stdlib.h>
#define FALSE 0
#define TRUE 1

typedef struct bstn {
    int value;
    struct bstn *left;
    struct bstn *right;
} BST;

/*@ pred BST(+x, K) {
  (x == NULL) * (K == -{ }-);

  x -m> struct bstn { int(#val); #left; #right } *
  (not (#left == NULL)) * (not (#right == NULL)) *
  BST(#right, #KR) * BST(#left, #KL) *
  (forall #z : Int. (#z --e-- #KL) => (#z <# #val)) *
  (forall #z : Int. (#z --e-- #KR) => (#val <# #z)) *
  (K == -u- (#KL, -{ #val }-, #KR));

  x -m> struct bstn { int(#val); NULL; #right } *
  (not (#right == NULL)) *
  BST(#right, #KR) *
  (forall #z : Int. #z --e-- #KR => #val <# #z) *
  (K == -u- (-{ #val }-, #KR));

  x -m> struct bstn { int(#val); #left; NULL } *
  (not (#left == NULL)) *
  BST(#left, #KL) *
  (forall #z : Int. #z --e-- #KL => #z <# #val) *
  (K == -u- (-{ #val }-, #KL));

  x -m> struct bstn { int(#val); NULL; NULL } *
  (K == -{ #val }-)
} */

/*@ spec makeNode(v) {
  requires: (v == #v) * (#v == int(#vv))
  ensures:  BST(ret, -{ #vv }-)
}*/
BST *makeNode(int v) {
    BST *new_node = malloc(sizeof(BST));
    new_node->value = v;
    new_node->left = NULL;
    new_node->right = NULL;
    return new_node;
}

/*@ spec insert(v, t) {
  requires: (v == int(#vv)) * (t == #t) * BST(#t, #K)
  ensures:  BST(ret, -u- (#K, -{ #vv }-))
}*/
BST *insert(int v, BST *t) {
    BST *tmp;
    if (t == NULL) {
        return makeNode(v);
    };
    if (v < t->value) {
        tmp = insert(v, t->left);
        __builtin_annot("unfold BST(tmp, -u- (-{ #vv }-, #KL))");
        t->left = tmp;
    } else if (v > t->value) {
        tmp = insert(v, t->right);
        __builtin_annot("unfold BST(tmp, -u- (-{ #vv }-, #KR))");
        t->right = tmp;
    };
    return t;
}

/*@ spec find(v, t) {
  requires: (v == int(#vv)) * (t == #t) * BST(#t, #K)
  ensures:  BST(#t, #K) * (#vv --e-- #K) * (ret == TRUE);
            BST(#t, #K) * (not (#vv --e-- #K)) * (ret == FALSE)
} */
int find(int v, BST *t) {
    if (t == NULL) {
        return FALSE;
    } else if (v == t->value) {
        return TRUE;
    } else if (v < t->value) {
        return find(v, t->left);
    } else { /* the only last case is v > t->value */
        return find(v, t->right);
    }
}

/*@ spec find_min(t) {
  requires: (t == #t) * BST(#t, #K) * (not (#t == NULL))
  ensures:  BST(#t, #K) * (ret == int(#r)) * (#r --e-- #K) *
            (forall #x : Int. (#x --e-- #K) => (#r <=# #x))
} */
int find_min(BST *t) {
    __builtin_annot("unfold BST(#t, #K)");
    __builtin_annot(
        "assert [[bind #left]] t -m> struct bstn { #a; #left; #right }");
    __builtin_annot("unfold BST(#left, #someSet)");
    if (t->left == NULL) {
        return t->value;
    } else {
        return find_min(t->left);
    }
}

/* spec remove(v, t) {
  requires: (v == int(#v)) * (t == #t) * BST(#t, #K)
  ensures:  (ret == #t_new) * BST(#t_new, #K_new) * (#K_new == #K -d- -{ #v }-)
} */
BST *remove(int v, BST *t) {
    BST *ret_node;
    if (t == NULL) {
        return NULL;
    } else if (v == t->value) {
        __builtin_annot("assert [[bind #left, #right]] t -m> struct bstn { "
                        "int(#v); #left; #right }");
        __builtin_annot("if (! (#left = NULL)) { unfold BST(#left, #KL) }");
        __builtin_annot("if (! (#right = NULL)) { unfold BST(#right, #KR) }");
        if (t->left == NULL) {
            ret_node = t->right;
            free(t);
            return ret_node;
        } else if (t->right == NULL) {
            ret_node = t->left;
            free(t);
            return ret_node;
        } else {
            int min = find_min(t->right);
            BST *trt = remove(min, t->right);
            __builtin_annot("unfold BST(trt, #NKR)");
            t->right = trt;
            t->value = min;
        }
    } else if (v < t->value) {
        BST *ntl = remove(v, t->left);
        __builtin_annot("unfold BST(ntl, #NKL)");
        t->left = ntl;
    } else {
        BST *ntr = remove(v, t->right);
        __builtin_annot("unfold BST(ntr, #NKR)");
        t->right = ntr;
    };
    return t;
}